### DISCUSSION TITLE
Run on GPU

### DISCUSSION THREAD

**Discussion by DanRRRR** [epochpic/epoch#705] [General]
Any news on implementing this functionality some time soon?

**Comment by DanRRRR** [comment]
Colleague compared 4 different PIC codes. Got that WarpX and EPOCH are the fastest. One of these codes was capable to run on NVIDIA GPU. Colleague  got the speedup of 1 GPU vs 1 CPU core... 173x !!!

**Reply by Status-Mirror** [reply]
Hey Dan,

This is still in development, but I wouldn't say it's coming out "soon". It's certainly something we're interested in pursuing - especially since GPU is the direction most modern clusters seem to be moving in. Hard to give a proper estimate for when this will be done, but it will be an important feature of the next EPOCH version.

Cheers,
Stuart

**Reply by DanRRRR** [reply]
Thanks for reply, Stuart. 
I'll just drop here couple words with this respect to urge and encourage you guys to have appropriate priorities set correctly.

Currently all supercomputer centers are substantially oversubscribed and the requests for computer time are very competitive. Major centers with the most powerful computers like Frontier, Aurora etc will simply not grant you any access if your code does not use GPU. 

And this is completely understandable. Each node of such computers like Frontier has one 64-core CPU chip and 8 GPU chips. Combined computing power of these 8 GPUs is  at least 10-20x  larger than from these  CPUs.  

In addition to that interesting to look  at what  the power consumption is in such computers per node. The CPU consume 300-400 W from the wall and 4 dual GPUs MI250 650W each or 2600W. That means GPU drain 6-8 times more than the CPU. Not using GPUs then becomes  just pure waste of resources because they continue to drain a lot even if not used. 

That was theoretical talk, most people do not care what is more powerful, what is less, and how many megawatts supercomputers drain, but all care about money. If compare processor cost with combined cost of GPU in one node (and there are 1e4 of them in Frontier) we will get that it is $5k vs $100k. Again waste of resources if not use GPU, nobody will allow that. 

And final shocking fact is that because typical request for CPU time for high end supercomputers like Frontier for example is almost 0.7-1 million node-hours, then let's do the math at 20 cents per kilowatt-hour -- only the cost of electricity for such request is 0.4-0.5 million dollars !! And 90% of that will go down to the drain if GPUs are not used

We are simply screwed and done if EPOCH will continue to be not GPU ready

**Reply by DanRRRR** [reply]
How things go with the GPU approach? 

If you are still busy with other nearby plans I've got to offer some unique I think opportunity to work on this. Our colleague we were working together in the past currently for last decade or so works on one of the world's top5  supercomputers and his recent duties among others were exactly to help supercomputer users to add GPU capabilities to their codes. If you  find  time from your side we all working together will test and hopefully quickly add GPU capabilities to EPOCH. Major supercomputers stop offering purely CPU runs from this year
